### PR TITLE
Support multi gpu in pytorch lm

### PR DESCRIPTION
**Pull request by ShigekiKarita**

### CODE REVIEW

File: egs/jsut/asr1/run.sh [hunk 1]
```diff
@@ -143,13 +143,6 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
...
     text2token.py -s 1 -n 1 data/${train_dev}/text | cut -f 2- -d" " \
         > ${lmdatadir}/valid.txt
-    # use only 1 gpu
-    if [ ${ngpu} -gt 1 ]; then
-        echo "LM training does not support multi-gpu. signle gpu will be used."
-        lmngpu=1
-    else
-        lmngpu=${ngpu}
-    fi
     ${cuda_cmd} --gpu ${lmngpu} ${lmexpdir}/train.log \
```
Comment (kan-bayashi) [inline egs/jsut/asr1/run.sh]:
```suggestion
    ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
```

File: egs/csj/asr1/run.sh [hunk 2]
```diff
@@ -168,13 +168,7 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
...
         > ${lmdatadir}/valid.txt
-    # use only 1 gpu
-    if [ ${ngpu} -gt 1 ]; then
-        echo "LM training does not support multi-gpu. signle gpu will be used."
-        lmngpu=1
-    else
-        lmngpu=${ngpu}
-    fi
+
     ${cuda_cmd} --gpu ${lmngpu} ${lmexpdir}/train.log \
```
Comment (kan-bayashi) [inline egs/csj/asr1/run.sh]:
```suggestion
    ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
```

File: egs/jsut/asr1/run.sh [hunk 3]
```diff
@@ -143,14 +143,7 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
...
         > ${lmdatadir}/valid.txt
-    # use only 1 gpu
-    if [ ${ngpu} -gt 1 ]; then
-        echo "LM training does not support multi-gpu. signle gpu will be used."
-        lmngpu=1
-    else
-        lmngpu=${ngpu}
-    fi
-    ${cuda_cmd} --gpu ${lmngpu} ${lmexpdir}/train.log \
+     ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
```
Comment (kan-bayashi) [inline egs/jsut/asr1/run.sh]:
```suggestion
    ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
```

File: egs/csj/asr1/run.sh [hunk 4]
```diff
@@ -168,14 +168,8 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
...
-    # use only 1 gpu
-    if [ ${ngpu} -gt 1 ]; then
-        echo "LM training does not support multi-gpu. signle gpu will be used."
-        lmngpu=1
-    else
-        lmngpu=${ngpu}
-    fi
-    ${cuda_cmd} --gpu ${lmngpu} ${lmexpdir}/train.log \
+
+     ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
```
Comment (kan-bayashi) [inline egs/csj/asr1/run.sh]:
```suggestion
    ${cuda_cmd} --gpu ${ngpu} ${lmexpdir}/train.log \
```

File: espnet/lm/pytorch_backend/lm.py [hunk 5]
```diff
@@ -350,21 +350,20 @@ def train(args):
...
     logging.info('#iterations per epoch = ' + str(len(train_iter.batch_indices)))
     logging.info('#total iterations = ' + str(args.epoch * len(train_iter.batch_indices)))
     # Prepare an RNNLM model
     rnn = RNNLM(args.n_vocab, args.layer, args.unit, args.type, args.dropout_rate)
     model = ClassifierWithState(rnn)
-    if args.ngpu > 1:
-        logging.warning("currently, multi-gpu is not supported. use single gpu.")
     if args.ngpu > 0:
-        # Make the specified GPU current
+        model = torch.nn.DataParallel(model).cuda()
```
Comment (kan-bayashi) [inline espnet/lm/pytorch_backend/lm.py]: I'm not sure `device_ids` is needed but it is used in `asr` and `tts`.
```suggestion
        model = torch.nn.DataParallel(model, device_ids=list(range(args.ngpu))).cuda()
```

File: espnet/lm/pytorch_backend/lm.py [hunk 6]
```diff
@@ -350,21 +350,20 @@ def train(args):
 
     use_sortagrad = args.sortagrad == -1 or args.sortagrad > 0
     # Create the dataset iterators
-    train_iter = ParallelSentenceIterator(train, args.batchsize,
+    batch_size = args.batchsize * max(args.ngpu, 1)
```
Comment (kan-bayashi) [inline espnet/lm/pytorch_backend/lm.py]: It is better to show info as the same as ASR.
https://github.com/espnet/espnet/blob/595cce96e1c5f2b71fa58f29deef0827a21d6ea8/espnet/asr/pytorch_backend/asr.py#L365-L366

**Comment by ShigekiKarita** [comment]
let's fix this after someone raises an issue

**Comment by ShigekiKarita** [comment]
I changed my mind. I will follow your comments to keep consistency.

**Pull request merged**